### PR TITLE
Fix embed blocks logic

### PR DESCRIPTION
**Pull request by toptalo**
I think that I fix issue #352

### CODE REVIEW

**Comment by boschr** [comment]
@toptalo why is this PR closed? This solves also my issue with nested embeds. Like:

```
{% embed '_visual.twig' %}
  {% block visual_left %}
    {% embed '_content-block.twig' with { title: 1 } only %}
    {% endembed %}
  {% endblock %}

  {% block visual_right %}
    {% embed '_content-block.twig' with { title: 2 } only %}
    {% endembed %}
  {% endblock %}
{% endembed %}```

**Comment by toptalo** [comment]
@boschr because @willrowe working on complete re-implementation of how blocks are handled

https://github.com/twigjs/twig.js/issues/287#issuecomment-376200875

**Comment by boschr** [comment]
Thnx for the quick reply. I saw the comment of @willrowe. But in my opinion we can still do some parallel fixes for the current solution.

**Comment by toptalo** [comment]
I can reopen this PR, and maintainers will be decide what to do with it

**Comment by boschr** [comment]
Sounds like a plan. Probably they wil merge it sooner when you add an test. Really hope that we can merge fixes of this size fast (same for my own PR #556).

**Comment by toptalo** [comment]
@dave-irvine Hello! Can you help me to understand one moment?

Here is example from `test.embed.js` https://twigfiddle.com/1e50jc And it fails in twigfiddle.

I think code in `embed-include.twig` is wrong and embed blocks should not overwrite template blocks.

Am i right?

Also I found that empty block in `extender` template must clear `extendee` block
here is example: https://twigfiddle.com/vpprve
To inherit extendee block content we need to insert `{{ parent() }}` tag

**Comment by toptalo** [comment]
@willrowe is it ok if I copy-paste this two tests?
* `should support complex nested embeds`
* `should support multiple inheritance and embeds`

**Comment by willrowe** [comment]
Sure, use whatever you need.

File: test/test.embed.js [hunk 1]
```diff
@@ -105,4 +105,40 @@ describe("Twig.js Embed ->", function() {
...
+            data: '<{% block section_title %}section-title{% endblock %}><{% block section_content %}section-content{% endblock %}>',
+            id: 'section.twig'
+        });
+
+        twig({
+            allowInlineIncludes: true,
+            data: '{% extends "layout.twig" %}{% block body_header %}override-body-header{% endblock %}{% block body_content %}{% embed "section.twig" %}{% block section_content %}override-section-content{% endblock %}{% endembed %}{% endblock %}'
+        }).render().should.equal('<layout-header><override-body-header><section-title><override-section-content><layout-body-footer><base-footer>');
+    });
+});
```
Comment (willrowe) [inline test/test.embed.js]: Please make sure there is a newline at the end of this file.

File: src/twig.logic.js [hunk 2]
```diff
@@ -1148,17 +1148,21 @@ module.exports = function (Twig) {
                         }
                     }
 
+                    // store previous blocks
+                    that._blocks = Object.assign({}, that.blocks);
```
Comment (willrowe) [inline src/twig.logic.js]: Please use `Twig.lib.copy` to be consistent with the rest of the code.

File: src/twig.logic.js [hunk 3]
```diff
@@ -1148,17 +1148,21 @@ module.exports = function (Twig) {
...
                     return Twig.parseAsync.call(that, token.output, innerContext)
                     .then(function() {
                         // render tempalte with blocks defined in embed block
-                        return template.renderAsync(innerContext, {'blocks':that.blocks});
+                        return template.renderAsync(innerContext, {'blocks': that.blocks});
                     });
                 })
                 .then(function(output) {
+                    // restore previous blocks
+                    that.blocks = Object.assign({}, that._blocks);
```
Comment (willrowe) [inline src/twig.logic.js]: Please use `Twig.lib.copy` to be consistent with the rest of the code.

**Comment by willrowe** [comment]
This is good to go as a bugfix, if you could merge this @dave-irvine and tag v1.11.1 that would be great.

**Comment by toptalo** [comment]
@willrowe thank you!

**Pull request merged**